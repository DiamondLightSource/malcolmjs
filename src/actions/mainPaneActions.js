/**
 * Created by twi18192 on 25/08/15.
 */

import AppDispatcher from '../dispatcher/appDispatcher.js';
import appConstants  from '../constants/appConstants.js';

class MainPaneActions {
  constructor()
    {

    }

  toggleFooter1(item){
    AppDispatcher.handleViewAction({
      actionType:appConstants.FOOTER_TOGGLE,
      item: item
    })
  }

  toggleConfigPanel(item){
    AppDispatcher.handleViewAction({
      actionType:appConstants.CONFIG_TOGGLE,
      item: item
    })
  }
  toggleFavPanel(item){
    AppDispatcher.handleViewAction({
      actionType: appConstants.FAV_TOGGLE,
      item: item
    })
  }
}

const mainPaneActions = new MainPaneActions();

export default mainPaneActions;
