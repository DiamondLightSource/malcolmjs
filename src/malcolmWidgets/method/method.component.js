import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { withStyles } from '@material-ui/core/styles';
import Divider from '@material-ui/core/Divider';
import Typography from '@material-ui/core/Typography';
import Tooltip from '@material-ui/core/Tooltip';
import IconButton from '@material-ui/core/IconButton';
import AttributeAlarm, {
  AlarmStates,
} from '../attributeDetails/attributeAlarm/attributeAlarm.component';
import ButtonAction from '../buttonAction/buttonAction.component';

import GroupExpander from '../groupExpander/groupExpander.component';
import blockUtils from '../../malcolm/blockUtils';
import {
  malcolmSetFlag,
  malcolmPostAction,
} from '../../malcolm/malcolmActionCreators';
import {
  malcolmUpdateMethodInput,
  malcolmArchivePost,
} from '../../malcolm/actions/method.actions';

import {
  selectorFunction,
  getDefaultFromType,
} from '../attributeDetails/attributeSelector/attributeSelector.component';
import navigationActions from '../../malcolm/actions/navigation.actions';

const styles = () => ({
  div: {
    display: 'flex',
    paddingLeft: 5,
    paddingRight: 5,
    alignItems: 'center',
  },
  textName: {
    flexGrow: 1,
    textAlign: 'Left',
    marginLeft: 4,
    marginRight: 4,
  },
  controlContainer: {
    width: 150,
    padding: 2,
  },
  runButton: {
    flexGrow: 1,
    padding: 2,
  },
  missingAttribute: {
    color: 'red',
  },
  button: {
    width: '22px',
    height: '22px',
    '&:hover': {
      backgroundColor: 'transparent',
    },
  },
});

const buildIOComponent = (input, props, isOutput) => {
  const parameterMeta = input[1];
  const widgetTag = parameterMeta.tags.find(t => t.indexOf('widget:') !== -1);
  const flags = {
    isDisabled: props.methodPending || !input[1].writeable,
    isErrorState: props.methodErrored && input[1].writeable,
    isDirty: props.dirtyInputs[input[0]],
  };
  const updateStoreOnEveryValueChange = true;

  const valueMap = isOutput ? props.outputValues : props.inputValues;
  let inputValue;
  if (valueMap[input[0]] !== undefined) {
    inputValue = valueMap[input[0]];
  } else if (props.defaultValues[input[0]] !== undefined) {
    inputValue = props.defaultValues[input[0]];
  } else {
    inputValue = getDefaultFromType(input[1]);
    if (inputValue !== undefined) {
      props.updateInput(props.methodPath, input[0], inputValue);
    }
  }
  const submitHandler = (path, value) => {
    props.updateInput(path, input[0], value);
  };
  const setFlag = (path, flagName, isDirty) => {
    props.updateInput(path, input[0], { isDirty });
  };
  const subElement = isOutput ? `returns.${input[0]}` : `takes.${input[0]}`;
  if (widgetTag) {
    return selectorFunction(
      widgetTag,
      [...props.methodPath, subElement],
      inputValue,
      submitHandler,
      flags,
      setFlag,
      props.theme.palette.primary.light,
      parameterMeta,
      false,
      updateStoreOnEveryValueChange,
      props.methodParamClickHandler
    );
  }
  return selectorFunction('widget:undefined');
};

const MethodDetails = props => {
  if (
    (!props.inputs || !Object.keys(props.inputs).length) &&
    (!props.outputs || !Object.keys(props.outputs).length)
  ) {
    return (
      <div className={props.classes.div}>
        <Tooltip title={props.methodErrorMessage}>
          <IconButton
            className={props.classes.button}
            disableRipple
            onClick={() =>
              props.buttonClickHandler(props.blockName, props.attributeName)
            }
          >
            <AttributeAlarm alarmSeverity={props.methodAlarm} />
          </IconButton>
        </Tooltip>
        <div className={props.classes.controlContainer}>
          <ButtonAction
            method
            text={props.methodName}
            disabled={
              !props.writeable || props.methodAlarm === AlarmStates.PENDING
            }
            clickAction={() =>
              props.runMethod(props.methodPath, props.inputValues)
            }
          />
        </div>
      </div>
    );
  }
  return (
    <GroupExpander key={props.methodPath} groupName={props.methodName} expanded>
      <div>
        {Object.entries(props.inputs).map(input => (
          <div key={input[0]} className={props.classes.div}>
            <Tooltip title={input[1].description}>
              <IconButton
                className={props.classes.button}
                disableRipple
                onClick={() =>
                  props.buttonClickHandler(
                    props.blockName,
                    props.attributeName,
                    `takes.${input[0]}`
                  )
                }
              >
                <AttributeAlarm alarmSeverity={AlarmStates.NO_ALARM} />
              </IconButton>
            </Tooltip>
            <Typography className={props.classes.textName}>
              {input[1].label}
            </Typography>
            <div className={props.classes.controlContainer}>
              {buildIOComponent(input, props, false)}
            </div>
          </div>
        ))}
        <div className={props.classes.div}>
          <Tooltip title={props.methodErrorMessage}>
            <IconButton
              className={props.classes.button}
              disableRipple
              onClick={() =>
                props.buttonClickHandler(props.blockName, props.attributeName)
              }
            >
              <AttributeAlarm alarmSeverity={props.methodAlarm} />
            </IconButton>
          </Tooltip>
          <div className={props.classes.controlContainer}>
            <ButtonAction
              method
              text={props.methodName}
              disabled={
                !props.writeable || props.methodAlarm === AlarmStates.PENDING
              }
              clickAction={() =>
                props.runMethod(props.methodPath, props.inputValues)
              }
            />
          </div>
        </div>
        {Object.keys(props.outputValues).length !== 0 ? (
          <div>
            <Typography className={props.classes.textName}>
              Last Return:
            </Typography>
            <Divider />
            {Object.entries(props.outputs).map(output => (
              <div key={output[0]} className={props.classes.div}>
                <Tooltip title={output[1].description}>
                  <IconButton
                    className={props.classes.button}
                    disableRipple
                    onClick={() =>
                      props.buttonClickHandler(
                        props.blockName,
                        props.attributeName,
                        `returns.${output[0]}`
                      )
                    }
                  >
                    <AttributeAlarm alarmSeverity={AlarmStates.NO_ALARM} />
                  </IconButton>
                </Tooltip>
                <Typography className={props.classes.textName}>
                  {output[1].label}
                </Typography>
                <div className={props.classes.controlContainer}>
                  {buildIOComponent(output, props, true)}
                </div>
              </div>
            ))}
          </div>
        ) : null}
      </div>
    </GroupExpander>
  );
};

MethodDetails.propTypes = {
  blockName: PropTypes.string.isRequired,
  attributeName: PropTypes.string.isRequired,
  methodName: PropTypes.string.isRequired,
  methodAlarm: PropTypes.number.isRequired,
  methodErrorMessage: PropTypes.string.isRequired,
  methodPath: PropTypes.arrayOf(PropTypes.string).isRequired,
  inputs: PropTypes.shape({}).isRequired,
  inputValues: PropTypes.shape({}).isRequired,
  outputs: PropTypes.shape({}).isRequired,
  outputValues: PropTypes.shape({}).isRequired,
  runMethod: PropTypes.func.isRequired,
  classes: PropTypes.shape({
    div: PropTypes.string,
    textName: PropTypes.string,
    runButton: PropTypes.string,
    controlContainer: PropTypes.string,
    button: PropTypes.string,
  }).isRequired,
  buttonClickHandler: PropTypes.func.isRequired,
  writeable: PropTypes.bool.isRequired,
};

const EMPTY = '';
const EMPTY_ARRAY = [];
const EMPTY_OBJECT = {};
const mapStateToProps = (state, ownProps) => {
  let method;
  if (ownProps.attributeName && ownProps.blockName) {
    method = blockUtils.findAttribute(
      state.malcolm.blocks,
      ownProps.blockName,
      ownProps.attributeName
    );
  }

  let alarm = AlarmStates.NO_ALARM;
  alarm =
    method && method.calculated.errorState ? AlarmStates.MAJOR_ALARM : alarm;
  alarm = method && method.calculated.pending ? AlarmStates.PENDING : alarm;

  return {
    writeable: method ? method.raw.writeable : false,
    methodName: method ? method.raw.label : 'Not found',
    methodAlarm: alarm,
    methodPending: method ? method.calculated.pending : false,
    methodErrored: method ? method.calculated.errorState : false,
    methodErrorMessage:
      method && method.calculated.errorMessage
        ? method.calculated.errorMessage
        : EMPTY,
    methodPath: method ? method.calculated.path : EMPTY_ARRAY,
    inputs: method ? method.raw.takes.elements : EMPTY_OBJECT,
    inputValues: (method && method.calculated.inputs) || EMPTY_OBJECT,
    dirtyInputs: (method && method.calculated.dirtyInputs) || EMPTY_OBJECT,
    outputs: method ? method.raw.returns.elements : EMPTY_OBJECT,
    outputValues: (method && method.calculated.outputs) || EMPTY_OBJECT,
    required: method ? method.raw.required : EMPTY_OBJECT,
    defaultValues: method ? method.raw.defaults : EMPTY_OBJECT,
  };
};

export const mapDispatchToProps = dispatch => ({
  buttonClickHandler: (blockName, attributeName, subElement) => {
    dispatch(
      navigationActions.navigateToInfo(blockName, attributeName, subElement)
    );
  },
  runMethod: (path, inputs) => {
    dispatch(malcolmSetFlag(path, 'pending', true));
    dispatch(malcolmArchivePost(path, inputs));
    dispatch(malcolmPostAction(path, inputs));
  },
  updateInput: (path, inputName, inputValue) => {
    dispatch(malcolmUpdateMethodInput(path, inputName, inputValue));
  },
  methodParamClickHandler: path => {
    dispatch(navigationActions.navigateToSubElement(path[0], path[1], path[2]));
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(
  withStyles(styles, { withTheme: true })(MethodDetails)
);
