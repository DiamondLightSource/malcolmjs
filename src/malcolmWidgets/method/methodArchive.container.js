import * as React from 'react';
import PropTypes from 'prop-types';
import CircularBuffer from 'circular-buffer';
import Typography from '@material-ui/core/Typography';
import Loadable from 'react-loadable';
import WidgetTable from '../table/table.component';
// import MethodPlot from './methodPlot.component';
import TabbedPanel from '../../attributeView/tabbedMiddlePanel.component';

const noOp = () => {};

const LoadableMethodPlot = Loadable({
  loader: () => import('./methodPlot.component'),
  loading: () => <Typography>Loading...</Typography>,
});

const MethodArchive = props => {
  const timeStamps = props.methodArchive.timeStamp.toarray();
  const values = props.methodArchive.value.toarray();
  const dummyAttribute = {
    raw: {
      value: {
        timeStamp: timeStamps.map(stamp => stamp.localRunTime.toISOString()),
        alarm: props.methodArchive.alarmState.toarray(),
        value: values.map(value => value.runParameters[props.selectedParam[1]]),
      },
      meta: {
        elements: {
          alarm: {
            tags: ['info:alarm'],
            label: 'Method Alarm state',
          },
          timeStamp: {
            tags: ['widget:textupdate'],
            label:
              props.selectedParam[0] === 'takes'
                ? 'Time sent'
                : 'Time received',
          },
          value: {
            tags: ['widget:textupdate'],
            label: 'Parameter value',
          },
        },
      },
    },
    calculated: {},
  };
  const dummyArchive = {
    parent: props.methodArchive.parent,
    name: props.methodArchive.name,
    timeStamp: timeStamps.map(
      stamp =>
        props.selectedParam[0] === 'takes'
          ? stamp.localRunTime
          : stamp.localReturnTime
    ),
    alarmState: props.methodArchive.alarmState.toarray(),
    value: values.map(value => value.runParameters[props.selectedParam[1]]),
  };
  return (
    <TabbedPanel
      alwaysUpdate
      openPanels={props.openPanels}
      tabLabels={['Table', 'Plot']}
      defaultTab={props.defaultTab}
    >
      <WidgetTable
        attribute={dummyAttribute}
        hideInfo
        eventHandler={noOp}
        setFlag={noOp}
        addRow={noOp}
        infoClickHandler={noOp}
        rowClickHandler={noOp}
      />
      <LoadableMethodPlot
        attribute={dummyArchive}
        openPanels={props.openPanels}
      />
    </TabbedPanel>
  );
};

MethodArchive.propTypes = {
  selectedParam: PropTypes.arrayOf(PropTypes.string).isRequired,
  methodArchive: PropTypes.shape({
    parent: PropTypes.string,
    name: PropTypes.string,
    timeStamp: PropTypes.instanceOf(CircularBuffer),
    value: PropTypes.instanceOf(CircularBuffer),
    alarmState: PropTypes.instanceOf(CircularBuffer),
  }).isRequired,
  openPanels: PropTypes.shape({
    parent: PropTypes.bool,
    child: PropTypes.bool,
  }).isRequired,
  defaultTab: PropTypes.number,
};
MethodArchive.defaultProps = {
  defaultTab: 0,
};

export default MethodArchive;
