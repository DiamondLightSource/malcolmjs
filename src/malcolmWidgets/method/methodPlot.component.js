import React from 'react';
import PropTypes from 'prop-types';
import { withTheme } from '@material-ui/core/styles';
import Plot from 'react-plotly.js';
import { AlarmStates } from '../attributeDetails/attributeAlarm/attributeAlarm.component';

const alarmStatesByIndex = [
  AlarmStates.NO_ALARM,
  AlarmStates.MINOR_ALARM,
  AlarmStates.MAJOR_ALARM,
  AlarmStates.UNDEFINED_ALARM,
];

const updatePlotData = (oldDataElement, alarmIndex, attribute) => {
  const dataElement = oldDataElement;
  const alarms = attribute.alarmState.toarray();
  dataElement.x = attribute.timeStamp.toarray();
  dataElement.y = attribute.Value.toarray().map(
    (value, valIndex) =>
      alarms[valIndex] === alarmStatesByIndex[alarmIndex] ||
      (alarms[valIndex - 1] === alarmStatesByIndex[alarmIndex] &&
        alarms[valIndex] !== AlarmStates.UNDEFINED_ALARM)
        ? value
        : null
  );
  dataElement.x.push(dataElement.x.slice(-1)[0]);
  dataElement.y.push(dataElement.y.slice(-1)[0]);
  dataElement.visible = dataElement.y.some(val => val !== null);
  return dataElement;
};

class MethodPlot extends React.Component {
  static getDerivedStateFromProps(props, state) {
    const { data, layout } = state;
    const dataRevision = `${props.attribute.parent}#${props.attribute.name}#${
      props.attribute.plotTime
    }#p:${props.openPanels.parent}#c:${props.openPanels.child}`;
    const newData = data.map((dataElement, index) =>
      updatePlotData(dataElement, index, props.attribute)
    );
    layout.datarevision = dataRevision;
    if (
      props.attribute.parent !== layout.parent ||
      props.attribute.name !== layout.attribute
    ) {
      layout.parent = props.attribute.parent;
      layout.attribute = props.attribute.name;
      layout.yaxis = {
        color: props.theme.palette.text.primary,
      };
    }
    return {
      ...state,
      data: [...newData],
      layout,
    };
  }

  constructor(props) {
    super(props);
    this.state = {
      data: [
        {
          x: [],
          y: [],
          type: 'scatter',
          mode: 'lines+points',
          marker: { color: props.theme.palette.primary.light },
          line: { shape: 'hv' },
          name: 'Normal',
          visible: false,
        },
        {
          x: [],
          y: [],
          type: 'scatter',
          mode: 'lines+points',
          marker: {
            color: props.theme.alarmState.warning,
          },
          line: { shape: 'hv', dash: '15px, 5px' },
          name: 'Minor',
          visible: false,
        },
        {
          x: [],
          y: [],
          type: 'scatter',
          mode: 'lines+points',
          marker: { color: props.theme.alarmState.error },
          line: { shape: 'hv', dash: '10px, 10px' },
          name: 'Major',
          visible: false,
        },
        {
          x: [],
          y: [],
          type: 'scatter',
          mode: 'lines+points',
          marker: { color: props.theme.alarmState.disconnected },
          line: { shape: 'hv', dash: '5px, 15px' },
          name: 'Disconnected',
          visible: false,
        },
      ],
      layout: {
        parent: '',
        attribute: '',
        legend: { font: { color: props.theme.palette.text.primary } },
        datarevision: -1,
        autosize: true,
        paper_bgcolor: props.theme.palette.background.default,
        plot_bgcolor: props.theme.palette.background.default,
        xaxis: {
          color: props.theme.palette.text.primary,
        },
        yaxis: {
          color: props.theme.palette.text.primary,
        },
      },
    };
  }

  render() {
    return (
      <Plot
        data={this.state.data}
        layout={this.state.layout}
        style={{ width: '100%', height: '100%' }}
        useResizeHandler
      />
    );
  }
}

MethodPlot.propTypes = {
  theme: PropTypes.shape({
    palette: PropTypes.shape({
      text: PropTypes.shape({
        primary: PropTypes.string,
      }),
      primary: PropTypes.shape({
        light: PropTypes.string,
      }),
      background: PropTypes.shape({
        default: PropTypes.string,
      }),
    }),
    alarmState: PropTypes.shape({
      warning: PropTypes.string,
      error: PropTypes.string,
      disconnected: PropTypes.string,
    }),
  }).isRequired,
  /*
  openPanels: PropTypes.shape({
    parent: PropTypes.bool,
    child: PropTypes.bool,
  }).isRequired,
  attribute: PropTypes.shape({
    parent: PropTypes.string,
    name: PropTypes.string,
    meta: PropTypes.shape({
      choices: PropTypes.arrayOf(PropTypes.string),
    }),
    value: PropTypes.arrayOf(PropTypes.number),
    timeStamp: PropTypes.arrayOf(PropTypes.number),
    plotTime: PropTypes.number,
    isBool: PropTypes.bool,
  }).isRequired,
  */
};

export default withTheme()(MethodPlot);
