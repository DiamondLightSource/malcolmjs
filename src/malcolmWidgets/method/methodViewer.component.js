/* eslint react/no-array-index-key: 0 */
import * as React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { withTheme } from '@material-ui/core/styles';
import JSONInput from 'react-json-editor-ajrm';
import locale from 'react-json-editor-ajrm/dist/locale/en';
// below is for updated version of json editor
// import locale from 'react-json-editor-ajrm/locale/en';
import Table from '@material-ui/core/Table';
import TableRow from '@material-ui/core/TableRow';
import TableFooter from '@material-ui/core/TableFooter';
import TableCell from '@material-ui/core/TableCell';
// import Typography from '@material-ui/core/Typography';
import ButtonAction from '../buttonAction/buttonAction.component';

import blockUtils from '../../malcolm/blockUtils';
import { malcolmUpdateMethodInput } from '../../malcolm/actions/method.actions';

const MethodViewer = props => {
  if (props.method && props.selectedParam) {
    const widgetTag = props.selectedParamMeta.tags.find(
      t => t.indexOf('widget:') !== -1
    );
    const transitionWithPanelStyle = {
      left: props.openParent ? 360 : 0,
      width: `calc(100% - ${(props.openChild ? 360 : 0) +
        (props.openParent ? 360 : 0)}px)`,
      // transition: 'width 1s, left 1s',
    };
    const { localStorage } = window;
    const footerItems = [
      ...props.footerItems,
      <ButtonAction
        text="Save to cookie"
        clickAction={() => {
          localStorage.setItem(
            `${props.blockName},${props.attributeName}.${props.selectedParam}`,
            JSON.stringify(props.selectedParamValue)
          );
        }}
      />,
      <ButtonAction
        text="Load from cookie"
        clickAction={() => {
          const savedVal = JSON.parse(
            localStorage.getItem(
              `${props.blockName},${props.attributeName}.${props.selectedParam}`
            )
          );
          props.updateInput(
            props.method.calculated.path,
            props.selectedParam[1],
            savedVal
          );
        }}
      />,
    ];
    switch (widgetTag) {
      case 'widget:tree':
        return (
          <div className={props.classes.plainBackground}>
            <div
              className={props.classes.tableContainer}
              style={{
                ...transitionWithPanelStyle,
                textAlign: 'left',
                display: 'initial',
              }}
            >
              <div style={{ height: 'calc(100% - 55px)' }}>
                <JSONInput
                  locale={locale}
                  placeholder={props.selectedParamValue}
                  onChange={val => {
                    if (!val.error) {
                      props.updateInput(
                        props.method.calculated.path,
                        props.selectedParam[1],
                        val.jsObject
                      );
                    }
                  }}
                  id={props.selectedParam}
                  height="100%"
                  width="100%"
                  style={{ body: { fontSize: '150%' } }}
                />
              </div>
              <Table>
                <TableFooter>
                  <TableRow>
                    {footerItems.map((item, key) => (
                      <TableCell key={key}>{item}</TableCell>
                    ))}
                  </TableRow>
                </TableFooter>
              </Table>
            </div>
          </div>
        );
      default:
        return <div className={props.classes.plainBackground} />;
    }
  }

  return <div>oops!</div>;
};

const mapStateToProps = (state, ownProps) => {
  let method;
  let selectedParamMeta;
  let selectedParamValue;
  if (ownProps.attributeName && ownProps.blockName) {
    method = blockUtils.findAttribute(
      state.malcolm.blocks,
      ownProps.blockName,
      ownProps.attributeName
    );
  }
  const selectedParam = ownProps.subElement;

  if (method && selectedParam) {
    selectedParamMeta = method.raw[selectedParam[0]].elements[selectedParam[1]];
    let ioType;
    if (selectedParam[0] === 'takes') {
      ioType = 'inputs';
    } else if (selectedParam[0] === 'returns') {
      ioType = 'outputs';
    }
    selectedParamValue = method.calculated[ioType]
      ? method.calculated[ioType][selectedParam[1]]
      : null;
  }
  return {
    method,
    selectedParam,
    selectedParamMeta,
    selectedParamValue,
  };
};

const mapDispatchToProps = dispatch => ({
  updateInput: (path, inputName, inputValue) => {
    dispatch(malcolmUpdateMethodInput(path, inputName, inputValue));
  },
});

MethodViewer.propTypes = {
  blockName: PropTypes.string.isRequired,
  attributeName: PropTypes.string.isRequired,
  method: PropTypes.shape({
    calculated: PropTypes.shape({
      path: PropTypes.arrayOf(PropTypes.string),
    }).isRequired,
    raw: PropTypes.shape({
      timeStamp: PropTypes.shape({
        secondsPastEpoch: PropTypes.string,
      }),
    }),
  }).isRequired,
  selectedParamMeta: PropTypes.shape({
    tags: PropTypes.arrayOf(PropTypes.string),
  }).isRequired,
  selectedParamValue: PropTypes.oneOf(
    PropTypes.string,
    PropTypes.number,
    PropTypes.object
  ).isRequired,
  classes: PropTypes.shape({
    plainBackground: PropTypes.string,
    tableContainer: PropTypes.string,
  }).isRequired,
  footerItems: PropTypes.arrayOf(PropTypes.node).isRequired,
  selectedParam: PropTypes.arrayOf(PropTypes.string).isRequired,
  openParent: PropTypes.bool.isRequired,
  openChild: PropTypes.bool.isRequired,
  updateInput: PropTypes.func.isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(
  withTheme()(MethodViewer)
);
