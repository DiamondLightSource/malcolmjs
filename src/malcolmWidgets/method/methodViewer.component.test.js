import React from 'react';
import { createShallow, createMount } from '@material-ui/core/test-utils';
import MethodViewer from './methodViewer.component';
import { malcolmUpdateMethodInput } from '../../malcolm/actions/method.actions';

jest.mock('../../malcolm/actions/method.actions');

describe('Method viewer', () => {
  let shallow;
  let mockStore;
  let mount;
  let testInputs;

  const getItem = jest.fn();
  const setItem = jest.fn();

  const state = {
    malcolm: {
      blocks: { Test: { attributes: [] } },
      blockArchive: { Test: { attributes: [] } },
    },
  };

  const testOutputs = {
    first: {
      label: 'First output',
      tags: ['widget:led'],
    },
    second: {
      label: 'Second output',
      tags: ['widget:textupdate'],
    },
  };

  const testInputValues = {
    first: 3,
    second: { value: 'four', isAString: true },
  };

  const testMethodViewer = (
    inputs,
    outputs,
    inputValues,
    selected,
    errorMsg
  ) => {
    const method = {
      calculated: {
        name: 'Method',
        path: ['Test', 'Method'],
        errorState: !!errorMsg,
        errorMessage: errorMsg,
        inputs: inputValues,
        outputs: {},
      },
      raw: {
        label: 'Test',
        defaults: {},
        takes: { elements: { ...inputs } },
        returns: { elements: { ...outputs } },
      },
    };

    state.malcolm.blocks.Test.attributes = [method];

    return (
      <MethodViewer
        store={mockStore(state)}
        blockName="Test"
        attributeName="Method"
        subElement={selected}
        classes={{}}
      />
    );
  };

  beforeEach(() => {
    getItem.mockClear();
    setItem.mockClear();
    malcolmUpdateMethodInput.mockClear();

    mockStore = myState => ({
      getState: () => myState,
      subscribe: () => {},
      dispatch: () => {},
    });

    shallow = createShallow({ dive: true });
    mount = createMount();

    testInputs = {
      first: {
        label: 'First input',
        description: 'a test input',
        tags: ['widget:textinput'],
      },
      second: {
        label: 'Second input',
        description: 'a test input',
        tags: ['widget:tree'],
      },
    };
  });

  afterEach(() => {
    mount.cleanUp();
  });

  it('renders correctly with no initial value', () => {
    const wrapper = shallow(
      testMethodViewer(testInputs, testOutputs, {}, ['takes', 'second'])
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('renders correctly with initial value', () => {
    const wrapper = shallow(
      testMethodViewer(testInputs, testOutputs, testInputValues, [
        'takes',
        'second',
      ])
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('renders empty div if selected parameter isnt a tree', () => {
    const wrapper = shallow(
      testMethodViewer(testInputs, testOutputs, {}, ['takes', 'first'])
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });

  // it('updates method input after 1s inactivity', () => {});
  /*
  it('save button hooks up correctly', () => {
    Object.defineProperty(window, 'localStorage', {
      value: { setItem, getItem },
    });
    const wrapper = mount(
      testMethodViewer(testInputs, testOutputs, testInputValues, [
        'takes',
        'second',
      ])
    );
    wrapper
      .find('button')
      .first()
      .simulate('click');
    expect(setItem).toHaveBeenCalledTimes(1);
    expect(setItem).toHaveBeenCalledWith(
      'Test,Method.takes,second',
      '{"value":"four","isAString":true}'
    );
  });

  it('load button hooks up correctly', () => {
    getItem.mockReturnValue('{"value":5,"isAString":false}');
    Object.defineProperty(window, 'localStorage', {
      value: { setItem, getItem },
    });
    const wrapper = mount(
      testMethodViewer(testInputs, testOutputs, {}, ['takes', 'second'])
    );
    wrapper
      .find('button')
      .last()
      .simulate('click');
    expect(getItem).toHaveBeenCalledTimes(1);
    expect(getItem).toHaveBeenCalledWith('Test,Method.takes,second');
    expect(malcolmUpdateMethodInput).toHaveBeenCalledTimes(1);
    expect(malcolmUpdateMethodInput).toHaveBeenCalledWith(
      ['Test', 'Method'],
      'second',
      {
        value: 5,
        isAString: false,
      }
    );
  }); */
});
