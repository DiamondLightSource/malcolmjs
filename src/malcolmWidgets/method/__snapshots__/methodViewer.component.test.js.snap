// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Method viewer renders correctly with initial value 1`] = `
<div>
  <div
    style={
      Object {
        "display": "initial",
        "left": 0,
        "textAlign": "left",
        "width": "calc(100% - 0px)",
      }
    }
  >
    <div
      style={
        Object {
          "height": "calc(100% - 56px)",
        }
      }
    >
      <t
        height="100%"
        id={
          Array [
            "takes",
            "second",
          ]
        }
        locale={
          Object {
            "brace": Object {
              "curly": Object {
                "cannotWrap": "'{token}' token cannot be wrapped in '{}' curly braces",
                "missingClose": "Open '{' curly brace is missing closing '}' curly brace",
                "missingOpen": "Missing '{' open curly brace",
              },
              "square": Object {
                "cannotWrap": "'{token}' token cannot be wrapped in '[]' square braces",
                "missingClose": "Open '[' square brace is missing closing ']' square brace",
                "missingOpen": "Missing '[' open square brace",
              },
            },
            "format": "{reason} at line {line}",
            "invalidToken": Object {
              "double": "'{token}' token cannot be followed by another '{token}' token",
              "sequence": "'{firstToken}' token cannot be followed by a '{secondToken}' token",
              "unexpected": "Unexpected '{token}' token found",
              "useInstead": "'{badToken}' token is not accepted. Use '{goodToken}' instead",
              "whitelist": "'{firstToken}' token can only follow '{secondToken}' token",
            },
            "key": Object {
              "numberAndLetterMissingQuotes": "Key beginning with number and containing letters must be wrapped by quotes",
              "spaceMissingQuotes": "Key containing space must be wrapped by quotes",
              "unexpectedString": "Unexpected string found at key position",
            },
            "noTrailingOrLeadingComma": "Trailing or leading commas in arrays and objects are not permitted",
            "string": Object {
              "missingClose": "Missing/invalid closing string '{quote}' token",
              "missingOpen": "Missing/invalid opening string '{quote}' token",
              "mustBeWrappedByQuotes": "Strings must be wrapped by quotes",
              "nonAlphanumeric": "Non-alphanemeric token '{token}' is not allowed outside string notation",
              "unexpectedKey": "Unexpected key found at string position",
            },
          }
        }
        onChange={[Function]}
        placeholder={
          Object {
            "isAString": true,
            "value": "four",
          }
        }
        style={
          Object {
            "body": Object {
              "fontSize": "150%",
            },
          }
        }
        width="100%"
      />
    </div>
    <WithStyles(Table)>
      <WithStyles(TableFooter)>
        <WithStyles(TableRow)>
          <WithStyles(TableCell)
            key="0"
          >
            <WithStyles(ButtonAction)
              clickAction={[Function]}
              text="Save to cookie"
            />
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="1"
          >
            <WithStyles(ButtonAction)
              clickAction={[Function]}
              text="Load from cookie"
            />
          </WithStyles(TableCell)>
        </WithStyles(TableRow)>
      </WithStyles(TableFooter)>
    </WithStyles(Table)>
  </div>
</div>
`;

exports[`Method viewer renders correctly with no initial value 1`] = `
<div>
  <div
    style={
      Object {
        "display": "initial",
        "left": 0,
        "textAlign": "left",
        "width": "calc(100% - 0px)",
      }
    }
  >
    <div
      style={
        Object {
          "height": "calc(100% - 56px)",
        }
      }
    >
      <t
        height="100%"
        id={
          Array [
            "takes",
            "second",
          ]
        }
        locale={
          Object {
            "brace": Object {
              "curly": Object {
                "cannotWrap": "'{token}' token cannot be wrapped in '{}' curly braces",
                "missingClose": "Open '{' curly brace is missing closing '}' curly brace",
                "missingOpen": "Missing '{' open curly brace",
              },
              "square": Object {
                "cannotWrap": "'{token}' token cannot be wrapped in '[]' square braces",
                "missingClose": "Open '[' square brace is missing closing ']' square brace",
                "missingOpen": "Missing '[' open square brace",
              },
            },
            "format": "{reason} at line {line}",
            "invalidToken": Object {
              "double": "'{token}' token cannot be followed by another '{token}' token",
              "sequence": "'{firstToken}' token cannot be followed by a '{secondToken}' token",
              "unexpected": "Unexpected '{token}' token found",
              "useInstead": "'{badToken}' token is not accepted. Use '{goodToken}' instead",
              "whitelist": "'{firstToken}' token can only follow '{secondToken}' token",
            },
            "key": Object {
              "numberAndLetterMissingQuotes": "Key beginning with number and containing letters must be wrapped by quotes",
              "spaceMissingQuotes": "Key containing space must be wrapped by quotes",
              "unexpectedString": "Unexpected string found at key position",
            },
            "noTrailingOrLeadingComma": "Trailing or leading commas in arrays and objects are not permitted",
            "string": Object {
              "missingClose": "Missing/invalid closing string '{quote}' token",
              "missingOpen": "Missing/invalid opening string '{quote}' token",
              "mustBeWrappedByQuotes": "Strings must be wrapped by quotes",
              "nonAlphanumeric": "Non-alphanemeric token '{token}' is not allowed outside string notation",
              "unexpectedKey": "Unexpected key found at string position",
            },
          }
        }
        onChange={[Function]}
        style={
          Object {
            "body": Object {
              "fontSize": "150%",
            },
          }
        }
        width="100%"
      />
    </div>
    <WithStyles(Table)>
      <WithStyles(TableFooter)>
        <WithStyles(TableRow)>
          <WithStyles(TableCell)
            key="0"
          >
            <WithStyles(ButtonAction)
              clickAction={[Function]}
              text="Save to cookie"
            />
          </WithStyles(TableCell)>
          <WithStyles(TableCell)
            key="1"
          >
            <WithStyles(ButtonAction)
              clickAction={[Function]}
              text="Load from cookie"
            />
          </WithStyles(TableCell)>
        </WithStyles(TableRow)>
      </WithStyles(TableFooter)>
    </WithStyles(Table)>
  </div>
</div>
`;

exports[`Method viewer renders empty div if selected parameter isnt a tree 1`] = `<div />`;
